File: frontend/src/utils/reconciliationAnalytics.ts
export interface AnalyticsRow {
  id: number
  dealer_id: number
  dealer_code: string
  dealer_name: string
  order_no: string
  part_no: string
  unit: string
  ordered_qty: string
  opening_qty: string
  closing_qty: string
  undated_qty: string
  daily: Record<string, string>
  baseline_remaining_qty?: string
}
export interface AnalyticsReport {
  start: string
  end: string
  dates: string[]
  generated_at: string
  actual: AnalyticsRow[]
  simulation: {
    available: boolean
    cutoff: string
    start: string
    end: string
    version: string
    reason: string
    imported?: boolean
    rows: AnalyticsRow[]
    events: Array<{ id: string; order_line_id: number; date: string; quantity: string; simulated: true }>
  }
}

// Integer hundredths avoid rounding drift when summing Decimal API strings.
export const cents = (value: string | number) => Math.round(Number(value) * 100)
export const sumQty = (rows: AnalyticsRow[], get: (row: AnalyticsRow) => string | number) => rows.reduce((sum, row) => sum + cents(get(row)), 0) / 100
export const formatQty = (value: string | number) => Number(value).toLocaleString('zh-CN', { maximumFractionDigits: 2 })
export const chartColors = ['#665cf6', '#208d8a', '#bd7623', '#6084c0', '#ae5f88']
export function groupOrders(rows: AnalyticsRow[]) {
  const groups = new Map<string, AnalyticsRow[]>()
  for (const row of rows) {
    const key = `${row.dealer_id}:${row.order_no}`
    groups.set(key, [...(groups.get(key) ?? []), row])
  }
  return [...groups].map(([key, items]) => {
    const ordered = sumQty(items, r => r.ordered_qty), remaining = sumQty(items, r => r.closing_qty)
    const invalid = items.some(r => Number(r.ordered_qty) <= 0 || Number(r.closing_qty) < 0 || Number(r.closing_qty) > Number(r.ordered_qty))
    const status = invalid ? '待核查' : items.every(r => cents(r.closing_qty) === 0) ? '已核销完' : items.every(r => cents(r.ordered_qty) === cents(r.closing_qty)) ? '未核销' : '部分核销'
    return { key, order: items[0]!.order_no, dealer: items[0]!.dealer_name, dealerId: items[0]!.dealer_id, ordered, remaining, status }
  })
}
export function dashboardRanges(day: string) {
  const last = new Date(`${day}T00:00:00Z`)
  last.setUTCDate(last.getUTCDate() - 29)
  return [
    { id: 'batch', label: '8.27—9.3', start: '2026-08-27', end: '2026-09-03' },
    { id: 'month', label: '本月', start: `${day.slice(0, 7)}-01`, end: day },
    { id: 'recent', label: '近 30 天', start: last.toISOString().slice(0, 10), end: day },
    { id: 'august', label: '8 月', start: '2026-08-01', end: '2026-08-31' },
  ]
}
export function summarize(rows: AnalyticsRow[], dates: string[]) {
  const daily = dates.map(date => ({ date, quantity: sumQty(rows, r => r.daily[date] ?? 0),
    lines: rows.filter(r => cents(r.daily[date] ?? 0) !== 0).length }))
  let balance = cents(sumQty(rows, r => r.opening_qty))
  const trend = daily.map(day => { balance -= cents(day.quantity); return { ...day, balance: balance / 100 } })
  return {
    daily: trend,
    orders: new Set(rows.map(r => `${r.dealer_id}:${r.order_no}`)).size,
    lines: rows.length,
    openLines: rows.filter(r => cents(r.closing_qty) > 0).length,
    closedLines: rows.filter(r => cents(r.closing_qty) === 0).length,
    opening: sumQty(rows, r => r.opening_qty),
    closing: sumQty(rows, r => r.closing_qty),
    quantity: daily.reduce((sum, day) => sum + cents(day.quantity), 0) / 100,
    activeDays: daily.filter(day => day.lines > 0).length,
  }
}

File: backend/app/services/reconciliation_analytics.py
"""Read-only charts and a reproducible, isolated historical demonstration.

Simulation never creates requests, ledger entries or outbound mail. Its opening
balances come from the committed import snapshot, not today's mutable balances.
"""
import hashlib
import json
from collections import defaultdict
from datetime import date, datetime, timedelta, timezone
from decimal import Decimal, ROUND_DOWN
from statistics import median

from fastapi import HTTPException
from sqlalchemy import case, func, select
from sqlalchemy.orm import Session, joinedload

from ..models import ImportJob, LedgerEntry, LedgerType, SalesOrderLine, ShipmentLine, ShipmentRequest, User, UserRole

CUTOFF = date(2026, 8, 26)
SIM_END = date(2026, 9, 3)
IMPORT_TOKEN = "5a6fa68cf8fc63fd9a9d86b033c890012bab040b8dcb1027"
PROMOTION_TOKEN = "approved-20260827-20260903-v1"
EXCLUDED_ORDERS = {"200018171"}
ZERO = Decimal(0)


def days_between(start: date, end: date) -> list[date]:
    return [start + timedelta(days=i) for i in range((end - start).days + 1)]


def amount(value: Decimal) -> str:
    return format(value.quantize(Decimal("0.01")), "f")


def identity(order: SalesOrderLine) -> str:
    return f"{order.dealer.code}|{order.order_no.strip()}|{order.part_no}"


def simulate_history(snapshot: list[dict], history: dict, visible_keys: set[str]) -> list[dict]:
    """Conservative scenario v1: six weekdays, three dispatch days per dealer.

    Group by unit, never pool unlike quantities. Up to two lines per dispatch;
    at most 15% of each line's opening remainder and 40 units per dispatch.
    Whole-period budget <=8% of remainder and <=45% of recent 28-day volume.
    Inactive dealers get <=1% of remainder, <=40 units (assumed small restart).
    """
    groups = defaultdict(list)
    for row in snapshot:
        if row["key"] in visible_keys and row["order_no"].strip() not in EXCLUDED_ORDERS:
            groups[(row["dealer"], row["unit"])].append(row)
    weekdays = [d for d in days_between(CUTOFF + timedelta(days=1), SIM_END) if d.weekday() < 5]
    events = []
    for (dealer, unit), rows in sorted(groups.items()):
        remaining = {r["key"]: Decimal(str(r["remaining_qty"])) for r in rows}
        opening = sum(remaining.values(), ZERO)
        recent = sum((q for r in rows for d, q in history.get(r["key"], {}).items()
                      if CUTOFF - timedelta(days=27) <= d <= CUTOFF and q > 0), ZERO)
        budget = min(opening * Decimal("0.08"), recent * Decimal("0.45")) if recent else min(opening * Decimal("0.01"), Decimal(40))
        budget = budget.quantize(Decimal(1), rounding=ROUND_DOWN)
        counts = defaultdict(int)
        offset = int(hashlib.sha256(dealer.encode()).hexdigest()[:8], 16) % 2
        for day in weekdays[offset::2]:
            def priority(row):
                last = max(history.get(row["key"], {}), default=date.min)
                tie = hashlib.sha256(f"scenario-v1|{day}|{row['key']}".encode()).hexdigest()
                return (counts[row["key"]], -last.toordinal(), tie)
            eligible = sorted((r for r in rows if remaining[r["key"]] >= 1 and counts[r["key"]] < 2), key=priority)
            for row in eligible[:2]:
                key = row["key"]
                samples = [q for d, q in history.get(key, {}).items() if q > 0 and d <= CUTOFF]
                typical = Decimal(median(samples)) * Decimal("0.5") if samples else Decimal(10)
                qty = min(budget, remaining[key], Decimal(str(row["remaining_qty"])) * Decimal("0.15"), typical, Decimal(40))
                qty = qty.quantize(Decimal(1), rounding=ROUND_DOWN)
                if qty <= 0:
                    continue
                events.append({"id": f"SIM-{day:%Y%m%d}-{hashlib.sha256(key.encode()).hexdigest()[:10]}",
                               "key": key, "date": day.isoformat(), "quantity": amount(qty), "simulated": True})
                budget -= qty
                remaining[key] -= qty
                counts[key] += 1
    return sorted(events, key=lambda e: (e["date"], e["key"]))


def reconciliation_analytics(db: Session, user: User, start: date, end: date) -> dict:
    if start > end or (end - start).days > 61 or start.year < 1900 or end.year > 2100:
        raise HTTPException(422, "请选择顺序正确且不超过 62 天的日期范围（1900—2100 年）")
    scope = SalesOrderLine.dealer_id == user.dealer_id if user.role == UserRole.DEALER else True
    orders = db.scalars(select(SalesOrderLine).where(scope).options(joinedload(SalesOrderLine.dealer))
                        .order_by(SalesOrderLine.dealer_id, SalesOrderLine.order_no, SalesOrderLine.part_no)).all()
    signed = case((LedgerEntry.entry_type == LedgerType.DEBIT, LedgerEntry.quantity), else_=-LedgerEntry.quantity)
    query = (select(LedgerEntry.order_line_id, ShipmentLine.requested_ship_date, func.sum(signed))
             .join(SalesOrderLine, LedgerEntry.order_line_id == SalesOrderLine.id)
             .join(ShipmentLine, LedgerEntry.shipment_line_id == ShipmentLine.id)
             .join(ShipmentRequest, ShipmentLine.request_id == ShipmentRequest.id)
             .where(scope, ShipmentRequest.dealer_id == SalesOrderLine.dealer_id)
             .group_by(LedgerEntry.order_line_id, ShipmentLine.requested_ship_date))
    daily = defaultdict(dict)
    for order_id, day, quantity in db.execute(query):
        daily[order_id][day] = Decimal(quantity)

    def row_data(order, ordered, opening, closing, cells, undated=ZERO):
        return dict(id=order.id, dealer_id=order.dealer_id, dealer_code=order.dealer.code,
                    dealer_name=order.dealer.name, order_no=order.order_no, part_no=order.part_no,
                    unit=order.unit, ordered_qty=amount(ordered), opening_qty=amount(opening),
                    closing_qty=amount(closing), undated_qty=amount(undated),
                    daily={d.isoformat(): amount(q) for d, q in cells.items() if start <= d <= end})

    actual_rows = []
    for order in orders:
        current = order.ordered_qty - order.reconciled_qty
        opening = current + sum((q for d, q in daily[order.id].items() if d >= start), ZERO)
        closing = current + sum((q for d, q in daily[order.id].items() if d > end), ZERO)
        undated = order.reconciled_qty - sum(daily[order.id].values(), ZERO)
        actual_rows.append(row_data(order, order.ordered_qty, opening, closing, daily[order.id], undated))

    promoted = db.scalar(select(ImportJob.committed_at).where(ImportJob.token == PROMOTION_TOKEN))
    if promoted:
        # Retire the scenario once approved: legacy clients cannot count it again.
        return dict(start=start.isoformat(), end=end.isoformat(), dates=[d.isoformat() for d in days_between(start, end)],
                    generated_at=datetime.now(timezone.utc).isoformat(), actual=actual_rows,
                    simulation=dict(available=False, imported=True, cutoff=CUTOFF.isoformat(), start="2026-08-27",
                                    end=SIM_END.isoformat(), version="conservative-v1", rows=[], events=[],
                                    reason="这批数据已获用户确认并纳入正式核销，请查看正式台账"))

    job = db.scalar(select(ImportJob).where(ImportJob.token == IMPORT_TOKEN, ImportJob.committed_at.is_not(None)))
    snapshot = json.loads(job.payload_json) if job else []
    visible = {identity(o): o for o in orders}
    # Confirmed unit corrections apply to both views; quantities stay frozen.
    snapshot = [{**r, "unit": visible[r["key"]].unit} if r["key"] in visible else r for r in snapshot]
    # Only the original manual-import DEBITs inform the historical scenario.
    # Later corrections remain in the real view, not in this frozen scenario.
    historical = defaultdict(dict)
    imported = (select(LedgerEntry.order_line_id, ShipmentLine.requested_ship_date, func.sum(LedgerEntry.quantity))
                .join(SalesOrderLine, LedgerEntry.order_line_id == SalesOrderLine.id)
                .join(ShipmentLine, LedgerEntry.shipment_line_id == ShipmentLine.id)
                .join(ShipmentRequest, ShipmentLine.request_id == ShipmentRequest.id)
                .where(scope, ShipmentRequest.dealer_id == SalesOrderLine.dealer_id,
                       ShipmentRequest.uid_validity == "manual-history", LedgerEntry.entry_type == LedgerType.DEBIT,
                       LedgerEntry.reason.contains("real-data-20260903"), ShipmentLine.requested_ship_date <= CUTOFF)
                .group_by(LedgerEntry.order_line_id, ShipmentLine.requested_ship_date))
    by_id = {o.id: identity(o) for o in orders}
    for order_id, day, quantity in db.execute(imported):
        historical[by_id[order_id]][day] = Decimal(quantity)
    events = simulate_history(snapshot, historical, set(visible))
    simulated = defaultdict(dict)
    for event in events:
        simulated[event["key"]][date.fromisoformat(event["date"])] = Decimal(event["quantity"])
    scenario_rows = []
    for record in snapshot:
        key = record["key"]
        if key not in visible or record["order_no"].strip() in EXCLUDED_ORDERS:
            continue
        order = visible[key]
        cells = {**historical[key], **simulated[key]}
        final = Decimal(str(record["remaining_qty"])) - sum(simulated[key].values(), ZERO)
        opening = final + sum((q for d, q in cells.items() if d >= start), ZERO)
        closing = final + sum((q for d, q in cells.items() if d > end), ZERO)
        row = row_data(order, Decimal(str(record["ordered_qty"])), opening, closing, cells)
        row["unit"] = record["unit"]
        row["baseline_remaining_qty"] = amount(Decimal(str(record["remaining_qty"])))
        scenario_rows.append(row)
    return dict(start=start.isoformat(), end=end.isoformat(), dates=[d.isoformat() for d in days_between(start, end)],
                generated_at=datetime.now(timezone.utc).isoformat(), actual=actual_rows,
                simulation=dict(available=bool(scenario_rows), cutoff=CUTOFF.isoformat(), start="2026-08-27", end=SIM_END.isoformat(),
                                version="conservative-v1", rows=scenario_rows,
                                events=[dict(id=e["id"], order_line_id=visible[e["key"]].id, date=e["date"], quantity=e["quantity"], simulated=True) for e in events],
                                reason="" if scenario_rows else "当前账号没有本次历史导入的订单，无法生成这批模拟数据"))
